library/VTKExtensions/Rendering/vtkF3DHexagonalBokehBlurPass.cxx

```diff
@@ -0,0 +1,387 @@
+#include "vtkF3DHexagonalBokehBlurPass.h"
+
+#include "vtkObjectFactory.h"
+#include "vtkOpenGLError.h"
+#include "vtkOpenGLFramebufferObject.h"
+#include "vtkOpenGLQuadHelper.h"
+#include "vtkOpenGLRenderUtilities.h"
+#include "vtkOpenGLRenderWindow.h"
+#include "vtkOpenGLRenderer.h"
+#include "vtkOpenGLShaderCache.h"
+#include "vtkOpenGLState.h"
+#include "vtkOpenGLVertexArrayObject.h"
+#include "vtkRenderState.h"
+#include "vtkRenderer.h"
+#include "vtkShaderProgram.h"
+#include "vtkTextureObject.h"
+
+vtkStandardNewMacro(vtkF3DHexagonalBokehBlurPass);
+
+constexpr std::string_view BlurFunc()
+{
+  // clang-format off
+  return R"#(
+const float PI = 3.14159265359;
+
+vec3 BlurTexture(sampler2D tex, vec2 uv, vec2 direction)
+{
+  vec3 color = vec3(0.0);
+  float acc = 0.0;
+
+  // fix for the Y shape artifacts
+  uv += 0.5 * invViewDims * direction;
+
+  const float step = 0.1;
+
+  for (float i = 0.0; i < 1.0; i += step)
+  {
+    vec2 offset = i * coc * direction * invViewDims;
+    color += texture(tex, uv + offset).rgb;
+    acc += 1.0;
+  }
+
+  return color / acc;
+})#";
+  // clang-format on
+}
+
+//------------------------------------------------------------------------------
+vtkF3DHexagonalBokehBlurPass::vtkF3DHexagonalBokehBlurPass() = default;
+
+//------------------------------------------------------------------------------
+vtkF3DHexagonalBokehBlurPass::~vtkF3DHexagonalBokehBlurPass()
+{
+  if (this->FrameBufferObject != nullptr)
+  {
+    vtkErrorMacro(<< "FrameBufferObject should have been deleted in ReleaseGraphicsResources().");
+  }
+  if (this->VerticalBlurTexture != nullptr)
+  {
+    vtkErrorMacro(<< "VerticalBlur should have been deleted in ReleaseGraphicsResources().");
+  }
+  if (this->DiagonalBlurTexture != nullptr)
+  {
+    vtkErrorMacro(<< "DiagonalBlur should have been deleted in ReleaseGraphicsResources().");
+  }
+}
+
+//------------------------------------------------------------------------------
+void vtkF3DHexagonalBokehBlurPass::PrintSelf(ostream& os, vtkIndent indent)
+{
+  this->Superclass::PrintSelf(os, indent);
+  os << indent << "CircleOfConfusion: " << this->CircleOfConfusion << std::endl;
+}
+
+//------------------------------------------------------------------------------
+void vtkF3DHexagonalBokehBlurPass::InitializeGraphicsResources(
+  vtkOpenGLRenderWindow* renWin, int w, int h)
+{
+  if (this->BackgroundTexture == nullptr)
+  {
+    this->BackgroundTexture = vtkTextureObject::New();
```

Comment:
looks like smart pointer is not really useful in that case ?

Reply:
It can. I'll switch to that